### PR TITLE
ratelimit adaptations

### PR DESCRIPTION
**Pull request by sbasan**
Created `on_request_hook` for `vManageSession` and `vManageAuth` which allows to attach custom user function which will be executed right before request is sent to vmanage.
Hooks can be assigned to a class (will attach to every new instance) or can be assigned after instance is created.
Added `ratelimit` utility module which demonstrates how global ratelimiting using `on_request_hook` can be implemented:
```python
from vmngclient.utils.ratelimit import ratelimit, on_request_throttle
ratelimit(max_requests_per_minute=100)
vManageSession.on_request_hook = on_request_throttle
vManageAuth.on_request_hook = on_request_throttle
```

### CODE REVIEW

**Comment by tomaszwarda** [comment]
Maybe change it to requests per second and set default to 100 requests per second as this is the maximum vmanage can handle according to [documentation](https://developer.cisco.com/docs/sdwan/#!browsing-returned-results-sorting-results-filtering-results-and-rate-limits/rate-limits-on-results).
This way it will enable the maximum possible requests and not depend on varying number of request from second to second in the timespan of one minute.

File: vmngclient/tests/test_utils_ratelimit.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    DEFAULT_MAX_REQUESTS_PER_MINUTE,
+    max_requests_per_minute_setting,
+    on_request_throttle,
+    ratelimit,
+    ratelimiters,
+    throttle,
+)
+
+
+class TestTypedList(unittest.TestCase):
```
Comment (cicharka) [inline vmngclient/tests/test_utils_ratelimit.py:17]: TestTypedList?

Comment (sbasan) [reply vmngclient/tests/test_utils_ratelimit.py:17]: will fix

**Comment by sbasan** [comment]
> Maybe change it to requests per second and set default to 100 requests per second as this is the maximum vmanage can handle according to [documentation](https://developer.cisco.com/docs/sdwan/#!browsing-returned-results-sorting-results-filtering-results-and-rate-limits/rate-limits-on-results). This way it will enable the maximum possible requests and not depend on varying number of request from second to second in the timespan of one minute.

RateLimiter uses very primitive implementation just forcing min_period = 1/max_rate. We don't use any kind of time window to control request rate dynamically. I decided to keep requests per minute (unit doesn't matter)

**Review by cicharka — APPROVED** [review]

**Comment by sbasan** [comment]
move to backlog

**Pull request closed without merging**